File: clients/android/ChexParity.kt
// Parity check for the in-process Kotlin validator (Chex.kt).
//
// Runs a battery of schema/data cases through BOTH the pure-Kotlin validator and
// the real `chex` binary (the oracle), and asserts they agree on accept/reject —
// the same lockstep guarantee the web and Flutter clients have.
//
//   kotlinc clients/android/Chex.kt clients/android/ChexParity.kt -include-runtime -d /tmp/chex-android.jar
//   java -jar /tmp/chex-android.jar ./dist-bin/chex
//
// Exits 0 when every case matches, 1 otherwise. Kotlin stdlib + JDK only. The
// validator under test takes native Maps; this harness serializes them to JSON
// (with a tiny encoder) only to feed the binary oracle.

import java.io.File
import java.nio.file.Files
import kotlin.system.exitProcess

data class ParityCase(
    val name: String,
    val schema: Map<String, Any?>,
    val data: Map<String, Any?>,
    val valid: Boolean,
)

val cases = listOf(
    ParityCase("primitive pass", mapOf("age" to "^[0-9]+$"), mapOf("age" to 30), true),
    ParityCase("primitive fail", mapOf("age" to "^[0-9]+$"), mapOf("age" to "x"), false),
    ParityCase("boolean coercion", mapOf("active" to "^(true|false)$"), mapOf("active" to true), true),
    ParityCase("nullable absent", mapOf("nickname?" to "^[a-z]+$"), mapOf(), true),
    ParityCase("nullable present ok", mapOf("nickname?" to "^[a-z]+$"), mapOf("nickname" to "ada"), true),
    ParityCase("nullable present bad", mapOf("nickname?" to "^[a-z]+$"), mapOf("nickname" to "A1"), false),
    ParityCase("missing required", mapOf("age" to "^[0-9]+$"), mapOf(), false),
    ParityCase("unknown property", mapOf("age" to "^[0-9]+$"), mapOf("age" to 1, "extra" to "x"), false),
    ParityCase("nested object ok", mapOf("addr" to mapOf("city" to "^[A-Za-z]+$")), mapOf("addr" to mapOf("city" to "Lagos")), true),
    ParityCase("nested object bad", mapOf("addr" to mapOf("city" to "^[A-Za-z]+$")), mapOf("addr" to mapOf("city" to "L4")), false),
    ParityCase("object type mismatch", mapOf("addr" to mapOf("city" to "^[A-Za-z]+$")), mapOf("addr" to "x"), false),
    ParityCase("scalar array ok", mapOf("tags" to listOf("^[a-z]+$")), mapOf("tags" to listOf("bun", "web")), true),
    ParityCase("scalar array bad", mapOf("tags" to listOf("^[a-z]+$")), mapOf("tags" to listOf("bun", "W1")), false),
    ParityCase("array type mismatch", mapOf("tags" to listOf("^[a-z]+$")), mapOf("tags" to "nope"), false),
    ParityCase("array of objects ok", mapOf("items" to listOf(mapOf("sku" to "^[A-Z0-9-]+$", "gift?" to "^(true|false)$"))), mapOf("items" to listOf(mapOf("sku" to "AB-1"), mapOf("sku" to "CD-2", "gift" to true))), true),
    ParityCase("array of objects bad", mapOf("items" to listOf(mapOf("sku" to "^[A-Z0-9-]+$"))), mapOf("items" to listOf(mapOf("sku" to "ab-1"))), false),
    ParityCase("record ok", mapOf("meta" to mapOf("^[a-z_]+$" to "^.+$")), mapOf("meta" to mapOf("a_b" to "x")), true),
    ParityCase("record bad key", mapOf("meta" to mapOf("^[a-z_]+$" to "^.+$")), mapOf("meta" to mapOf("A" to "x")), false),
    ParityCase("record bad value", mapOf("meta" to mapOf("^[a-z]+$" to "^[0-9]+$")), mapOf("meta" to mapOf("a" to "x")), false),
)

// Minimal JSON encoder — only to feed the binary oracle.
fun toJson(value: Any?): String = when (value) {
    null -> "null"
    is String -> "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"") + "\""
    is Boolean, is Int, is Long, is Double -> value.toString()
    is Map<*, *> -> value.entries.joinToString(",", "{", "}") { "${toJson(it.key.toString())}:${toJson(it.value)}" }
    is List<*> -> value.joinToString(",", "[", "]") { toJson(it) }
    else -> throw IllegalArgumentException("unsupported JSON value: $value")
}

fun inProcessAccepts(schema: Map<String, Any?>, data: Map<String, Any?>): Boolean =
    try {
        CHEXValidator.validate(schema, data)
        true
    } catch (e: CHEXException) {
        false
    }

fun binaryAccepts(bin: String, dir: File, index: Int, schema: Map<String, Any?>, data: Map<String, Any?>): Boolean {
    val file = File(dir, "s$index.schema.json").apply { writeText(toJson(schema)) }
    val proc = ProcessBuilder(bin, "validate", file.path, toJson(data))
        .redirectOutput(ProcessBuilder.Redirect.DISCARD)
        .redirectError(ProcessBuilder.Redirect.DISCARD)
        .start()
    return proc.waitFor() == 0
}

fun main(args: Array<String>) {
    val bin = if (args.isNotEmpty()) args[0] else "chex"
    val dir = Files.createTempDirectory("chex-android-").toFile()
    var failures = 0
    try {
        cases.forEachIndexed { index, c ->
            val oracle = binaryAccepts(bin, dir, index, c.schema, c.data)
            val inProc = inProcessAccepts(c.schema, c.data)
            if (!(oracle == c.valid && inProc == oracle)) {
                failures++
                System.err.println("MISMATCH ${c.name}: expected=${c.valid} binary=$oracle inProcess=$inProc")
            }
        }
    } finally {
        dir.deleteRecursively()
    }
    if (failures == 0) {
        println("parity OK: ${cases.size} cases agree with the chex binary")
    } else {
        System.err.println("$failures/${cases.size} cases FAILED")
        exitProcess(1)
    }
}

File: clients/web/chex_parity.mjs
// Smoke test for the Wasm binding in chex.mjs.
//
// chex.mjs loads the same engine the binary compiles from, so the validation
// rules can't disagree — what this checks is the binding: the linear-memory
// copy, the allocator handshake, the result buffer, and error decoding. It runs
// a battery of cases through both and asserts they agree on accept/reject.
//
//   bun ./scripts/build-wasm.mjs
//   bun clients/web/chex_parity.mjs ./target/release/chex
//
// Exits 0 when every case matches, 1 otherwise.

import { mkdtemp, rm, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'

import { CHEXError, ready, validate } from './chex.mjs'

const cases = [
    { name: 'primitive pass', schema: { age: '^[0-9]+$' }, data: { age: 30 }, valid: true },
    { name: 'primitive fail', schema: { age: '^[0-9]+$' }, data: { age: 'x' }, valid: false },
    { name: 'boolean coercion', schema: { active: '^(true|false)$' }, data: { active: true }, valid: true },
    { name: 'nullable absent', schema: { 'nickname?': '^[a-z]+$' }, data: {}, valid: true },
    { name: 'nullable present ok', schema: { 'nickname?': '^[a-z]+$' }, data: { nickname: 'ada' }, valid: true },
    { name: 'nullable present bad', schema: { 'nickname?': '^[a-z]+$' }, data: { nickname: 'A1' }, valid: false },
    { name: 'missing required', schema: { age: '^[0-9]+$' }, data: {}, valid: false },
    { name: 'unknown property', schema: { age: '^[0-9]+$' }, data: { age: 1, extra: 'x' }, valid: false },
    { name: 'nested object ok', schema: { addr: { city: '^[A-Za-z]+$' } }, data: { addr: { city: 'Lagos' } }, valid: true },
    { name: 'nested object bad', schema: { addr: { city: '^[A-Za-z]+$' } }, data: { addr: { city: 'L4' } }, valid: false },
    { name: 'object type mismatch', schema: { addr: { city: '^[A-Za-z]+$' } }, data: { addr: 'x' }, valid: false },
    { name: 'scalar array ok', schema: { tags: ['^[a-z]+$'] }, data: { tags: ['bun', 'web'] }, valid: true },
    { name: 'scalar array bad', schema: { tags: ['^[a-z]+$'] }, data: { tags: ['bun', 'W1'] }, valid: false },
    { name: 'array type mismatch', schema: { tags: ['^[a-z]+$'] }, data: { tags: 'nope' }, valid: false },
    {
        name: 'array of objects ok',
        schema: { items: [{ sku: '^[A-Z0-9-]+$', 'gift?': '^(true|false)$' }] },
        data: { items: [{ sku: 'AB-1' }, { sku: 'CD-2', gift: true }] },
        valid: true,
    },
    {
        name: 'array of objects bad',
        schema: { items: [{ sku: '^[A-Z0-9-]+$' }] },
        data: { items: [{ sku: 'ab-1' }] },
        valid: false,
    },
    { name: 'record ok', schema: { meta: { '^[a-z_]+$': '^.+$' } }, data: { meta: { a_b: 'x' } }, valid: true },
    { name: 'record bad key', schema: { meta: { '^[a-z_]+$': '^.+$' } }, data: { meta: { A: 'x' } }, valid: false },
    { name: 'record bad value', schema: { meta: { '^[a-z]+$': '^[0-9]+$' } }, data: { meta: { a: 'x' } }, valid: false },
]

const binary = process.argv[2] ?? 'chex'
const wasm = process.env.CHEX_WASM ?? 'dist-web/chex.wasm'

// Compile from bytes rather than fetch(file://) so this runs headless.
await ready({ module: await WebAssembly.compile(await Bun.file(wasm).arrayBuffer()) })

const dir = await mkdtemp(join(tmpdir(), 'chex-web-'))
let failures = 0
let counter = 0

try {
    for (const testCase of cases) {
        const oracle = await binaryAccepts(testCase.schema, testCase.data)
        const viaWasm = wasmAccepts(testCase.schema, testCase.data)
        if (oracle !== testCase.valid || viaWasm !== oracle) {
            failures += 1
            console.error(
                `MISMATCH ${testCase.name}: expected=${testCase.valid} binary=${oracle} wasm=${viaWasm}`
            )
        }
    }

    // The contract the browser depends on: the original object back, and an
    // error carrying the engine's own class rather than a generic one.
    const data = { age: 30 }
    if (validate({ age: '^[0-9]+$' }, data) !== data) {
        failures += 1
        console.error('MISMATCH: validate did not return the original data object')
    }
    try {
        validate({ age: '^[0-9]+$' }, { age: 'x' }, 'user.schema.json')
        failures += 1
        console.error('MISMATCH: a failing validate did not throw')
    } catch (error) {
        if (!(error instanceof CHEXError) || error.name !== 'ValidationError') {
            failures += 1
            console.error(`MISMATCH: expected a ValidationError, got ${error.name}`)
        }
        if (!error.message.includes("in schema 'user.schema.json'")) {
            failures += 1
            console.error(`MISMATCH: the label is missing from ${error.message}`)
        }
    }
} finally {
    await rm(dir, { recursive: true, force: true })
}

if (failures > 0) {
    console.error(`${failures}/${cases.length} cases FAILED`)
    process.exit(1)
}
console.log(`parity OK: ${cases.length} cases agree with the chex binary`)

function wasmAccepts(schema, data) {
    try {
        validate(schema, data)
        return true
    } catch {
        return false
    }
}

async function binaryAccepts(schema, data) {
    const file = join(dir, `s${counter++}.schema.json`)
    await writeFile(file, JSON.stringify(schema))
    const proc = Bun.spawn([binary, 'validate', file, JSON.stringify(data)], {
        stdout: 'ignore',
        stderr: 'ignore',
        stdin: 'ignore',
    })
    return (await proc.exited) === 0
}
